File: Demos/Demo-OC/Demo-OC/XFDemoViewController.m
//
//  XFDemoViewController.m
//  Demo-OC
//
//  Created by 胡文峰 on 2018/12/18.
//  Copyright © 2018 XIAOFUTECH. All rights reserved.
//

#import "XFDemoViewController.h"

@interface XFDemoViewController ()

@end

@implementation XFDemoViewController

- (void)dealloc
{
    NSLog(@"【dealloc】%@", self);
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.

    self.navigationController.navigationBar.translucent = NO;

    self.title = @"XF Sdk Demo";
    self.view.backgroundColor = XFColor(0xf6f6f6, 1.0f);

    [self setupFuncs];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
 #pragma mark - Navigation

 // In a storyboard-based application, you will often want to do a little preparation before navigation
 - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
 // Get the new view controller using [segue destinationViewController].
 // Pass the selected object to the new view controller.
 }
 */

- (void)setupFuncs
{
    CGRect frame = CGRectMake(0, 0, 50, 30);
    UIButton *settingBtn = [[UIButton alloc] initWithFrame:frame];
    [settingBtn setTitleColor:XFColor(0x333333, 0.8f) forState:UIControlStateNormal];
    [settingBtn setTitle:@"设置" forState:UIControlStateNormal];
    settingBtn.titleLabel.font = XFFont(17);
    [settingBtn addTarget:self action:@selector(openApplicationSettings:)
         forControlEvents:UIControlEventTouchUpInside];
    UIBarButtonItem *item = [[UIBarButtonItem alloc] initWithCustomView:settingBtn];
    self.navigationItem.rightBarButtonItems = @[item];

    // ------------------------------------------------------------------------------------------

    CGFloat width = kWidth/3;
    CGFloat height = 56;

    frame = CGRectMake(width/3, 15, width, height);
    UIButton *cameraDrive = [self setupButtonTitle:@"CTBleHelper"
                                            Action:@selector(showCameraDrive) Frame:frame];
    [self.view addSubview:cameraDrive];

    frame = CGRectMake([cameraDrive xf_GetRight] + width/3, [cameraDrive xf_GetTop], width, height);
    UIButton *easyLinker = [self setupButtonTitle:@"CTEasyLinker\n（推荐）"
                                             Action:@selector(showEasyLinker) Frame:frame];
    [self.view addSubview:easyLinker];

    frame = CGRectMake([cameraDrive xf_GetRight] + width/3,
                       [cameraDrive xf_GetBottom] + width/3, width, height);
    UIButton *swiftLinker = [self setupButtonTitle:@"CTSwiftLinker\n（推荐+）"
                                           Action:@selector(showSwiftLinker) Frame:frame];
    [self.view addSubview:swiftLinker];
}

- (void)openApplicationSettings:(UIButton *)sender
{
    [XFAppDelegate XF_ApplicationOpenSettingsType:0];
}

- (UIButton *)setupButtonTitle:(NSString *)title Action:(SEL)action Frame:(CGRect)frame
{
    UIButton *button = [UIButton XF_ButtonWithColor:[UIColor clearColor] Image:nil
                                              Title:title Font:XFFont(15)
                                         TitleColor:XFColor(0xffffff, 1.0f) Target:self
                                             Action:action Frame:frame];
    button.userInteractionEnabled = NO;

    button.titleLabel.adjustsFontSizeToFitWidth = YES;
    button.layer.cornerRadius = 3.0f;
    button.clipsToBounds = YES;
    [button xf_SetBackgroundColor:XFColor(0xcecece, 1.0f) forState:UIControlStateNormal];
    [button xf_SetBackgroundColor:XFColor(0x4d7bfe, 1.0f) forState:UIControlStateSelected];
    [button xf_SetBackgroundColor:XFColor(0x4d7bfe, 0.75f)
                         forState:UIControlStateSelected | UIControlStateHighlighted];

    button.userInteractionEnabled = YES;
    button.selected = YES;
    button.titleLabel.numberOfLines = 0;
    button.titleLabel.textAlignment = NSTextAlignmentCenter;
    return button;
}

- (void)showCameraDrive
{
    UIViewController *ctr = [NSClassFromString(@"BleHelperViewController") new];
    [self.navigationController pushViewController:ctr animated:YES];
}

- (void)showEasyLinker
{
    UIViewController *ctr = [NSClassFromString(@"EasyLinkerViewController") new];
    [self.navigationController pushViewController:ctr animated:YES];
}

- (void)showSwiftLinker
{
    UIViewController *ctr = [NSClassFromString(@"SwiftLinkerViewController") new];
    [self.navigationController pushViewController:ctr animated:YES];
}

@end

File: Demos/Demo-Swift/Demo-Swift/XFDemoViewController.swift
//
//  XFDemoViewController.swift
//  Demo-Swift
//
//  Created by 胡文峰 on 2019/1/12.
//  Copyright © 2019 XIAOFUTECH. All rights reserved.
//

import UIKit

class XFDemoViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.

        self.navigationController?.navigationBar.isTranslucent = false
        self.title = "XF Sdk Demo"
        self.view.backgroundColor = UIColor.XF_f6Gray

        self.setupFuncs();
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning();
        // Dispose of any resources that can be recreated.
    }

    func setupFuncs() {
        var frame: CGRect = CGRect.init(x: 0, y: 0, width: 50, height: 30)
        let settingBtn: UIButton = UIButton.init(frame: frame)
        settingBtn.setTitle("设置", for: UIControl.State.normal)
        settingBtn.titleLabel?.font = XFFont(fontSize: 17.0)
        settingBtn.setTitleColor(XFColor(rgbValue: 0x333333, alpha: 1.0), for: UIControl.State.normal)
        settingBtn.addTarget(self, action: #selector(openApplicationSettings(sender:)), for: UIControl.Event.touchUpInside)
        let item = UIBarButtonItem.init(customView: settingBtn)
        self.navigationItem.rightBarButtonItems = [item]

        // ------------------------------------------------------------------------------------------

        let width: CGFloat = XFWidth()/3
        let height: CGFloat = 56.0

        frame = CGRect.init(x: width/3, y: 15, width: width, height: height)
        let cameraDrive: UIButton = self.setupButtonTitle(title: "CTBleHelper",
                                                          action: #selector(showCameraDrive), frame: frame)
        self.view.addSubview(cameraDrive)

        frame = CGRect.init(x: cameraDrive.xf_GetRight() + width/3, y: cameraDrive.xf_GetTop(),
                            width: width, height: height)
        let easyLinker: UIButton = self.setupButtonTitle(title: "CTEasyLinker\n（推荐）",
                                                         action: #selector(showEasyLinker),
                                                         frame: frame)
        self.view.addSubview(easyLinker)

        frame = CGRect.init(x: cameraDrive.xf_GetRight() + width/3,
                            y: cameraDrive.xf_GetBottom() + width/3,
                            width: width, height: height)
        let swiftLinker: UIButton = self.setupButtonTitle(title: "CTSwiftLinker\n（推荐+）",
                                                          action: #selector(showSwiftLinker),
                                                          frame: frame)
        self.view.addSubview(swiftLinker)
    }

    @objc func openApplicationSettings(sender: UIButton) {
        XF_ApplicationOpenSettings(type: 0)
    }

    func setupButtonTitle(title: NSString, action: Selector, frame: CGRect) -> UIButton {
        let button: UIButton = UIButton.init(type: UIButton.ButtonType.custom)
        button.frame = frame
        button.backgroundColor = UIColor.clear
        button.addTarget(self, action: action, for: UIControl.Event.touchUpInside)

        button.setTitle(title as String, for: UIControl.State.normal)
        button.titleLabel?.font = XFFont(fontSize: 15.0)
        button.titleLabel?.numberOfLines = 0
        button.titleLabel?.textAlignment = NSTextAlignment.center
        button.titleLabel?.adjustsFontSizeToFitWidth = true
        button.setTitleColor(XFColor(rgbValue: 0xffffff, alpha: 1.0), for: UIControl.State.normal)

        button.xf_SetBackgroundColor(XFColor(rgbValue: 0xcecece, alpha: 1.0), for: UIControl.State.normal)
        button.xf_SetBackgroundColor(XFColor(rgbValue: 0x4d7bfe, alpha: 1.0), for: UIControl.State.selected)
        button.xf_SetBackgroundColor(XFColor(rgbValue: 0x4d7bfe, alpha: 0.75),
                                     for: [UIControl.State.selected, UIControl.State.highlighted])

        button.layer.cornerRadius = 3.0
        button.clipsToBounds = true

        button.isSelected = true
        return button
    }

    @objc func showCameraDrive() {
        let ctr: UIViewController = BleHelperViewController()
        self.navigationController?.pushViewController(ctr, animated: true)
    }

    @objc func showEasyLinker() {
        let ctr: UIViewController = EasyLinkerViewController()
        self.navigationController?.pushViewController(ctr, animated: true)
    }

    @objc func showSwiftLinker() {
        //let nameSpace: NSString = Bundle.main.infoDictionary!["CFBundleExecutable"] as! NSString
        //let className = (nameSpace as String) + "." + "SwiftLinkerViewController"
        //let Clz = NSClassFromString(className) as! UIViewController.Type
        //let ctr: UIViewController = Clz.init()

        let ctr: UIViewController = SwiftLinkerViewController()
        self.navigationController?.pushViewController(ctr, animated: true)
    }

}
